curriculum/challenges/english/20-upcoming-python/learn-classes-and-objects-by-building-a-sudoku-solver/6569d83fe4dcc614c2ff971d.md

```diff
@@ -0,0 +1,81 @@
+---
+id: 6569d83fe4dcc614c2ff971d
+title: Step 41
+challengeType: 20
+dashedName: step-41
+---
+
+# --description--
+
+For each element in the specified column (`col`) of the current row (`row`), check whether the value at the current position in the 2D list is not equal to the provided `num`.
+
+# --hints--
+
+Access the element at the current position in the 2D list using `self.board[row][col]`.
+
+```js
+assert.match(code, /self\.board\s*\[\s*row\s*\]\s*\[\s*col\s*\]/);
+```
+
+Check if the element at the current position in the 2D list is not equal to the provided `num`.
+
+```js
+assert.match(code, /self\.board\s*\[\s*row\s*\]\s*\[\s*col\s*\]\s*!=\s*num/);
+```
+
+# --seed--
+
+## --seed-contents--
+
+```py
+class Board:
+    def __init__(self, board):
+        self.board = board
+
+    def __str__(self):
+        upper_lines = f'\n╔═══{"╤═══"*2}{"╦═══"}{"╤═══"*2}{"╦═══"}{"╤═══"*2}╗\n'
+        middle_lines = f'╟───{"┼───"*2}{"╫───"}{"┼───"*2}{"╫───"}{"┼───"*2}╢\n'
+        lower_lines = f'╚═══{"╧═══"*2}{"╩═══"}{"╧═══"*2}{"╩═══"}{"╧═══"*2}╝\n'
+        board_string = upper_lines
+        for index, line in enumerate(self.board):
+            row_list = []
+            for square_no, part in enumerate([line[:3], line[3:6], line[6:]], start=1):
+                row_square = '|'.join(str(item) for item in part)
+                row_list.extend(row_square)
+                if square_no != 3:
+                    row_list.append('║')
+
+            row = f'║ {" ".join(row_list)} ║\n'
+            row_empty = row.replace('0', ' ')
+            board_string += row_empty
+
+            if index < 8:
+                if index % 3 == 2:
+                    board_string += f'╠═══{"╪═══"*2}{"╬═══"}{"╪═══"*2}{"╬═══"}{"╪═══"*2}╣\n'
+                else:
+                    board_string += middle_lines
+            else:
+                board_string += lower_lines
+
+        return board_string
+
+    def find_empty_cell(self):
+        for row, contents in enumerate(self.board):
+            try:
+                col = contents.index(0)
+                return row, col
+            except ValueError:
+                pass
+        return None
+
+    def valid_in_row(self, row, num):
+        return num not in self.board[row]
+
+    
+--fcc-editable-region--
+    def valid_in_col(self, col, num):
+        
+        # change code above this line
+        for row in range(9)
```

Comment:
```suggestion
        for row in range(9):
```

Reply:
`:` wont be needed as the complete statement would be like this as it builds up gradually in the coming steps:

```py
return all(
            self.board[row][col] != num
            for row in range(9)
        )
```